mem/buffers.go

```diff
@@ -0,0 +1,156 @@
+// Free is invoked on all references.
+type Buffer struct {
+	data  []byte
+	refs  *atomic.Int32
+	free  func([]byte)
+	freed bool
+}
+
+// NewBuffer creates a new buffer from the given data, initializing the reference
+// counter to 1. Note that the given data is not copied.
+func NewBuffer(data []byte, free func([]byte)) *Buffer {
+	b := &Buffer{data: data, refs: new(atomic.Int32), free: free}
+	b.refs.Add(1)
+	return b
+}
+
+// Copy first invokes BufferPool.Get to acquire a new buffer of the size of the
+// given buffer, then copies the given buffer's data into the new buffer, and
+// finally returns a Buffer that, when freed, will be returned to the pool.
+func Copy(data []byte, pool BufferPool) *Buffer {
+	buf := pool.Get(len(data))
+	copy(buf, data)
+	return NewBuffer(buf, pool.Put)
+}
+
+// ReadOnlyData returns the underlying byte slice. Note that it is undefined
+// behavior to modify the contents of this slice in any way.
+func (b *Buffer) ReadOnlyData() []byte {
+	if b.freed {
+		panic("Cannot read freed buffer")
+	}
+	return b.data
+}
+
+// Ref returns a new reference to this Buffer's underlying buffer. Calling Free
+// on this reference will only free the buffer if all other references have been
+// freed.
+func (b *Buffer) Ref() *Buffer {
+	if b.freed {
+		panic("Cannot ref freed buffer")
+	}
+
+	b.refs.Add(1)
+	return &Buffer{
+		data: b.data,
+		refs: b.refs,
+		free: b.free,
+	}
+}
+
+// Free decrements this Buffer's reference counter. It then frees the underlying
+// buffer if the counter reaches 0 as a result of this call, does nothing
+// otherwise, or if Free has already been called. All subsequent calls to
+// ReadOnlyData will panic. Noop if receiver is nil.
+func (b *Buffer) Free() {
+	if b == nil || b.freed {
+		return
+	}
+
+	b.freed = true
+	refs := b.refs.Add(-1)
+	if refs == 0 && b.free != nil {
+		b.free(b.data)
+	}
+	b.data = nil
+}
+
+// Len returns the Buffer's size.
+func (b *Buffer) Len() int {
+	// Convenience: io.Reader returns (n int, err error), and n is often checked
+	// before err is checked. To mimic this, Len() should work on nil Buffers.
+	if b == nil {
+		return 0
+	}
+	return len(b.ReadOnlyData())
+}
+
+// Split modifies the receiver to point to the first n bytes while it returns a
+// new reference to the remaining bytes. The returned Buffer functions just like
+// a normal reference acquired using Ref in that the underlying buffer will only
+// be freed once both the receiver and the returned Buffer are freed.
+func (b *Buffer) Split(n int) *Buffer {
+	if b.freed {
+		panic("Cannot split freed buffer")
+	}
+
+	b.refs.Add(1)
+
+	data := b.data
+	var free func([]byte)
+	if f := b.free; f != nil {
+		free = func(_ []byte) {
+			f(data)
+		}
+	}
+
+	b.data = data[:n]
+	b.free = free
+
+	return &Buffer{
+		data: data[n:],
+		refs: b.refs,
+		free: free,
+	}
+}
+
+// String returns a string representation of the buffer that contains the address
```

Comment:
I would prefer for this docstring to simply say:
```
// String returns a string representation of the buffer. Maybe used for debugging purposes.
```
instead of describing exactly how it will look.

And also, we should avoid saying things like `used to uniquely identify the underlying buffer`, since very quickly our users will start relying on the output of this method for wildly critical things, and will complain when we change the implementation.